"panic: negative refcount" when mutating tags in starlark for metrics from a tail input

### Relevant telegraf.conf

```toml
[[outputs.discard]]

[[inputs.tail]]
  files = ["/var/log/something.log"]
  data_format = "json"
  json_time_key = "Time"
  json_time_format = "unix"
  watch_method = "poll"

[[processors.starlark]]
  script = "/etc/telegraf/postprocess.star"
```


### Logs from Telegraf

```text
panic: negative refcount

goroutine 53 [running]:
github.com/influxdata/telegraf/metric.(*trackingMetric).decr(0xc002608f00?)
/go/src/github.com/influxdata/telegraf/metric/tracking.go:145 +0x85
github.com/influxdata/telegraf/metric.(*trackingMetric).Drop(0xc0021b9180?)
/go/src/github.com/influxdata/telegraf/metric/tracking.go:139 +0x13
github.com/influxdata/telegraf/agent.(*Agent).runAggregators.func1()
/go/src/github.com/influxdata/telegraf/agent/agent.go:720 +0xee
created by github.com/influxdata/telegraf/agent.(*Agent).runAggregators in goroutine 114
/go/src/github.com/influxdata/telegraf/agent/agent.go:707 +0x329
```


### System info

Telegraf >= 1.29.0

### Docker

Here is a minimal way to reproduce the error:

- `docker-compose.yaml`

    ```yaml
    version: "3"
    
    services:
      telegraf:
        image: telegraf:1.29
        volumes:
          - ./telegraf.conf:/etc/telegraf/telegraf.conf:ro
          - ./postprocess.star:/etc/telegraf/postprocess.star:ro
          - ./something.log:/var/log/something.log:ro
        command: --debug
    ```

- `./telegraf.conf` is the configuration file posted above
- `./postprocess.star`

    ```python
    def apply(metric):
        # Cloning the metric before mutating the tags *appears* to solve the issue
        # Would this have any unforeseen consequences? Is this a valid workaround?
        # metric = deepcopy(metric)
    
        print(metric)
        metric.tags.pop("path", None)

        # Other operations that cause the panic include setting tag values:
        # metrics.tags["path"] = "some-override"
    
        return [metric]
    ```

- `./something.log` should start as an empty file


### Steps to reproduce

1. Run the the above setup with `docker compose up`
2. Write to `something.log` using something like

    ```sh
    echo '{ "Value": 1, "Time": "'(date +%s)'" }' >> something.log
    ```

3. You'll see the output of the `print` line in the `starlark` showing the metric being processed
4. Wait until the metrics are flushed to the output — at that point you'll get a `panic: negative refcount` message and the docker container will exit



### Expected behavior

- No golang panics
- Metric tags should be able to be mutated in the `starlark` processor


### Actual behavior

- golang panics when metric tags are mutated in `starlark`

### Additional info

- This bug is only present from 1.29 onwards (you can easily see this by changing the `image` tag in the `docker-compose.yaml` file shown above)
- Seems to be closely related to https://github.com/influxdata/telegraf/issues/13148
- There may be a bunch of other metric mutation operations that also cause this panic but we have only been mutating tags
- This only seems to happen for `tail` inputs — `file` inputs don't suffer from the same issue

--- a/plugins/common/starlark/metric.go
+++ b/plugins/common/starlark/metric.go
@@ -50,6 +50,9 @@ func (m *Metric) String() string {
 	buf.WriteString(", time=")
 	buf.WriteString(m.Time().String())
 	buf.WriteString(")")
+	if m.ID != 0 {
+		buf.WriteString(fmt.Sprintf("[tracking ID=%v]", m.ID))
+	}
 	return buf.String()
 }
 

--- a/plugins/common/starlark/metric.go
+++ b/plugins/common/starlark/metric.go
@@ -12,8 +12,7 @@ import (
 )
 
 type Metric struct {
-	ID uint64
-
+	ID             telegraf.TrackingID
 	metric         telegraf.Metric
 	tagIterCount   int
 	fieldIterCount int
@@ -22,7 +21,9 @@ type Metric struct {
 
 // Wrap updates the starlark.Metric to wrap a new telegraf.Metric.
 func (m *Metric) Wrap(metric telegraf.Metric) {
-	m.ID = metric.HashID()
+	if tm, ok := metric.(telegraf.TrackingMetric); ok {
+		m.ID = tm.TrackingID()
+	}
 	m.metric = metric
 	m.tagIterCount = 0
 	m.fieldIterCount = 0

--- a/plugins/common/starlark/metric.go
+++ b/plugins/common/starlark/metric.go
@@ -51,7 +51,7 @@ func (m *Metric) String() string {
 	buf.WriteString(m.Time().String())
 	buf.WriteString(")")
 	if m.ID != 0 {
-		buf.WriteString(fmt.Sprintf("[tracking ID=%v]", m.ID))
+		fmt.Fprintf(buf, "[tracking ID=%v]", m.ID)
 	}
 	return buf.String()
 }

--- a/plugins/processors/starlark/starlark.go
+++ b/plugins/processors/starlark/starlark.go
@@ -77,7 +77,7 @@ func (s *Starlark) Add(origMetric telegraf.Metric, acc telegraf.Accumulator) err
 
 				// Previous metric was found, accept the starlark metric, add
 				// the original metric to the accumulator
-				if v.ID == origMetric.HashID() {
+				if v.ID != 0 {
 					origFound = true
 					s.results = append(s.results, origMetric)
 					acc.AddMetric(origMetric)
@@ -106,7 +106,7 @@ func (s *Starlark) Add(origMetric telegraf.Metric, acc telegraf.Accumulator) err
 		m := rv.Unwrap()
 		// If we got the original metric back, use that and drop the new one.
 		// Otherwise mark the original as accepted and use the new metric.
-		if origMetric.HashID() == rv.ID {
+		if rv.ID != 0 {
 			acc.AddMetric(origMetric)
 		} else {
 			origMetric.Accept()

--- a/plugins/processors/starlark/starlark.go
+++ b/plugins/processors/starlark/starlark.go
@@ -54,13 +54,13 @@ func (s *Starlark) Add(origMetric telegraf.Metric, acc telegraf.Accumulator) err
 	}
 	parameters[0].(*common.Metric).Wrap(origMetric)
 
-	rv, err := s.Call("apply")
+	returnValue, err := s.Call("apply")
 	if err != nil {
 		s.LogError(err)
 		return err
 	}
 
-	switch rv := rv.(type) {
+	switch rv := returnValue.(type) {
 	case *starlark.List:
 		iter := rv.Iterate()
 		defer iter.Done()

--- a/plugins/processors/starlark/starlark_test.go
+++ b/plugins/processors/starlark/starlark_test.go
@@ -3477,6 +3477,92 @@ def apply(metric):
 	}
 }
 
+func TestTrackingStateful(t *testing.T) {
+	var testCases = []struct {
+		name     string
+		source   string
+		results  int
+		loops    int
+		delivery int
+	}{
+		{
+			name:     "delayed release",
+			loops:    4,
+			results:  3,
+			delivery: 4,
+			source: `
+state = {"last": None}
+
+def apply(metric):
+  previous = state["last"]
+  state["last"] = deepcopy(metric)
+  return previous
+`,
+		},
+		{
+			name:     "delayed release with tracking",
+			loops:    4,
+			results:  3,
+			delivery: 3,
+			source: `
+state = {"last": None}
+
+def apply(metric):
+  previous = state["last"]
+  state["last"] = deepcopy(metric, track=True)
+  return previous
+`,
+		},
+	}
+	for _, tt := range testCases {
+		t.Run(tt.name, func(t *testing.T) {
+			// Create a tracking metric and tap the delivery information
+			var mu sync.Mutex
+			delivered := make([]telegraf.TrackingID, 0, tt.delivery)
+			notify := func(di telegraf.DeliveryInfo) {
+				mu.Lock()
+				defer mu.Unlock()
+				delivered = append(delivered, di.ID())
+			}
+
+			// Configure the plugin
+			plugin := newStarlarkFromSource(tt.source)
+			require.NoError(t, plugin.Init())
+			acc := &testutil.Accumulator{}
+			require.NoError(t, plugin.Start(acc))
+
+			// Do the requested number of loops
+			expected := make([]telegraf.TrackingID, 0, tt.loops)
+			for i := 0; i < tt.loops; i++ {
+				// Process expected metrics and compare with resulting metrics
+				input, tid := metric.WithTracking(testutil.TestMetric(i), notify)
+				expected = append(expected, tid)
+				require.NoError(t, plugin.Add(input, acc))
+			}
+			plugin.Stop()
+			expected = expected[:tt.delivery]
+
+			// Simulate output acknowledging delivery of metrics and check delivery
+			actual := acc.GetTelegrafMetrics()
+			// Ensure we get back the correct number of metrics
+			require.Lenf(t, actual, tt.results, "expected %d metrics but got %d", tt.results, len(actual))
+			for _, m := range actual {
+				m.Accept()
+			}
+
+			require.Eventuallyf(t, func() bool {
+				mu.Lock()
+				defer mu.Unlock()
+				return len(delivered) >= tt.delivery
+			}, 1*time.Second, 100*time.Millisecond, "original metric(s) not delivered")
+
+			mu.Lock()
+			defer mu.Unlock()
+			require.ElementsMatch(t, expected, delivered, "mismatch in delivered metrics")
+		})
+	}
+}
+
 // parses metric lines out of line protocol following a header, with a trailing blank line
 func parseMetricsFrom(t *testing.T, lines []string, header string) (metrics []telegraf.Metric) {
 	parser := &influx.Parser{}

--- a/plugins/processors/starlark/starlark_test.go
+++ b/plugins/processors/starlark/starlark_test.go
@@ -3428,6 +3428,15 @@ def apply(metric):
     return [x]
 `,
 		},
+		{
+			name:       "issue #14484",
+			numMetrics: 1,
+			source: `
+def apply(metric):
+    metric.tags.pop("tag1")
+    return [metric]
+`,
+		},
 	}
 	for _, tt := range testCases {
 		t.Run(tt.name, func(t *testing.T) {

